kubert/src/runtime.rs

```diff
@@ -295,11 +384,57 @@ impl<S> Runtime<S> {
         T::DynamicType: Default,
     {
         let watch = watcher::watcher(api, watcher_config);
+        #[cfg(feature = "prometheus-client")]
+        let watch = self.instrument_watch(watch);
         let successful = errors::LogAndSleep::fixed_delay(self.error_delay, watch);
         let initialized = self.initialized.add_handle().release_on_ready(successful);
         shutdown::CancelOnShutdown::new(self.shutdown_rx.clone(), initialized)
     }
 
+    #[cfg(feature = "prometheus-client")]
+    fn instrument_watch<T>(
+        &self,
+        watch: impl Stream<Item = watcher::Result<watcher::Event<T>>> + Send,
+    ) -> impl Stream<Item = watcher::Result<watcher::Event<T>>> + Send
+    where
+        T: Resource + DeserializeOwned + Clone + Debug + Send + 'static,
+        T::DynamicType: Default,
+    {
+        let dt = Default::default();
+        let labels = ResourceWatchLabels {
+            kind: T::kind(&dt).to_string(),
+            group: T::group(&dt).to_string(),
+            version: T::version(&dt).to_string(),
+        };
+        let metrics = self.metrics.clone();
+
+        watch.map(move |event| {
+            if let Some(metrics) = &metrics {
+                match event {
+                    Ok(watcher::Event::Applied(_)) => {
+                        metrics.watch_applies.get_or_create(&labels).inc();
+                    }
+                    Ok(watcher::Event::Restarted(_)) => {
+                        metrics.watch_restarts.get_or_create(&labels).inc();
+                    }
+                    Ok(watcher::Event::Deleted(_)) => {
+                        metrics.watch_deletes.get_or_create(&labels).inc();
+                    }
+                    Err(ref e) => {
+                        let error_labels = ResourceWatchErrorLabels {
+                            kind: labels.kind.clone(),
+                            group: labels.group.clone(),
+                            version: labels.version.clone(),
+                            error: e.to_string(),
```

Comment:
I don't feel comfortable dumping arbitrary error strings into our metrics. These can be high cardinality and dependent on internal codepaths that change release to release. Don't we have useful error type information here?